### PR TITLE
[Azure][Storage] Update default storage account naming with subscription id hash

### LINKED ISSUE #3783: [Azure blob] Storage account already exists in a different subscription
I switched azure subscriptions recently. 

Tried launching a simple YAML in the new subscription:
```
file_mounts:
  /outputs:
    name: romil-az-test
    store: azure
    source: /Users/romilb/tmp-workdir/
    mode: MOUNT
```

But failed with:
```
(base) ➜  sky-experiments git:(fix_test_docker_storage) ✗  sky launch -c test task.yaml 
Task from YAML spec: task.yaml
I 07-24 17:35:25 storage.py:2260] Created Azure resource group 'sky2ea485ea'.
E 07-24 17:35:33 storage.py:901] Could not create StoreType.AZURE store with name romil-az-test.
Traceback (most recent call last):
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 2265, in _get_storage_account_and_resource_group
    self.storage_client.storage_accounts.get_properties(
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/azure/core/tracing/decorator.py", line 78, in wrapper_use_tracer
    return func(*args, **kwargs)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/azure/mgmt/storage/v2022_09_01/operations/_storage_accounts_operations.py", line 1071, in get_properties
    map_error(status_code=response.status_code, response=response, error_map=error_map)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/azure/core/exceptions.py", line 164, in map_error
    raise error
azure.core.exceptions.ResourceNotFoundError: (ResourceNotFound) The Resource 'Microsoft.Storage/storageAccounts/skyeastus2ea485ea' under resource group 'sky2ea485ea' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix
Code: ResourceNotFound
Message: The Resource 'Microsoft.Storage/storageAccounts/skyeastus2ea485ea' under resource group 'sky2ea485ea' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 2296, in _create_storage_account
    self.storage_client.storage_accounts.begin_create(
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/azure/core/tracing/decorator.py", line 78, in wrapper_use_tracer
    return func(*args, **kwargs)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/azure/mgmt/storage/v2022_09_01/operations/_storage_accounts_operations.py", line 913, in begin_create
    raw_result = self._create_initial(  # type: ignore
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/azure/mgmt/storage/v2022_09_01/operations/_storage_accounts_operations.py", line 768, in _create_initial
    map_error(status_code=response.status_code, response=response, error_map=error_map)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/azure/core/exceptions.py", line 164, in map_error
    raise error
azure.core.exceptions.ResourceExistsError: (StorageAccountAlreadyTaken) The storage account named skyeastus2ea485ea is already taken.
Code: StorageAccountAlreadyTaken
Message: The storage account named skyeastus2ea485ea is already taken.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 893, in add_store
    store = store_cls(
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 1993, in __init__
    super().__init__(name, source, region, is_sky_managed,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 261, in __init__
    self.initialize()
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 2142, in initialize
    self._get_storage_account_and_resource_group())
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 2271, in _get_storage_account_and_resource_group
    self._create_storage_account(resource_group_name,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 2315, in _create_storage_account
    raise exceptions.StorageBucketCreateError(
sky.exceptions.StorageBucketCreateError: Failed to create storage account 'skyeastus2ea485ea'. You may be attempting to create a storage account already being in use. Details: [azure.core.exceptions.ResourceExistsError] (StorageAccountAlreadyTaken) The storage account named skyeastus2ea485ea is already taken.
Code: StorageAccountAlreadyTaken
Message: The storage account named skyeastus2ea485ea is already taken.
```

We should have some kind of fallback here - is the storage account name is already taken we should add a random suffix and retry?

### PR DESCRIPTION
**Pull request by landscapepainter**
This resolves #3783. 

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] launching on azure blob storage by switching subscription IDs
  1. `az login` to subscription A
  2. run `sky launch azure_blob.yaml` with subscription A, which creates/uses default storage account #A
  3. `az logout` and `az login` to subscription B
  4. run `sky launch azure_blob.yaml` with subscription B, whic creates/uses default storage account #B
- [x] Backward compatibility tests:
  - [x] check if sky managed storages created under old default storage accounts can be fetched after updating to new default storage account name with subscription ID hash in the code base:
  1. create storage under old default storage account without Azure subscription ID hash.
  4. update code so that new default storage account uses Azure subscription ID hash.
      1. If used same storage name from i., then it successfully fetches the sky managed  container belonging to the old default storage account.
      2. If used different storage name from i., then it creates a new default storage account with subscription ID hash, and creates the container under it.

- [x] Smoke tests `pytest tests/test_smoke.py::TestStorageWithCredentials --azure`
- [x] `pytest tests/test_smoke.py::test_azure_storage_mounts_with_stop --azure`
- [x] `sky launch comp_test.yaml --cloud azure -y`

`comp_test.yaml`:
```
file_mounts:
  #### Azure Blob tests ####
  /az-sky-managed-copy:
    name: az-sky-managed-copy
    source: ~/sky_logs
    store: azure
    mode: COPY

  /az-sky-managed-mount:
    name: az-sky-managed-mount
    source: ~/sky_logs
    store: azure
    mode: MOUNT

  /az-external-private-by-user:
    source: https://mystorageaccount.blob.core.windows.net/az-external-private-by-user-dy
    mode: MOUNT

  /az-external-public-by-user:
    source: https://mystorageaccount.blob.core.windows.net/az-external-public-by-user-dy
    mode: MOUNT

  /az-external-public-not-by-user:
    source: https://azureopendatastorage.blob.core.windows.net/nyctlc
    mode: MOUNT

  #### S3 tests ####
  /s3-sky-managed-copy:
    name: s3-sky-managed-copy
    source: ~/sky_logs
    store: s3
    mode: COPY

  /s3-sky-managed-mount:
    name: s3-sky-managed-mount
    source: ~/sky_logs
    store: s3
    mode: MOUNT

  /s3-external-private-by-user:
    source: s3://s3-external-private-by-user-dy
    mode: MOUNT

  /s3-external-public-by-user:
    source: s3://s3-external-public-by-user-dy
    mode: MOUNT

  /s3-external-public-not-by-user:
    source: s3://digitalcorpora
    mode: MOUNT

  #### GCS tests ####
  /gcs-sky-managed-copy:
    name: gcs-sky-managed-copy
    source: ~/sky_logs
    store: gcs
    mode: COPY

  /gcs-sky-managed-mount:
    name: gcs-sky-managed-mount
    source: ~/sky_logs
    store: gcs
    mode: MOUNT

  /gcs-external-private-by-user:
    source: gs://gcs-external-private-by-user-dy
    mode: MOUNT

  /gcs-external-public-by-user:
    source: gs://gcs-external-public-by-user-dy
    mode: MOUNT

  /gcs-external-public-not-by-user:
    source: gs://gcp-public-data-sentinel-2
    mode: MOUNT

workdir: ~/yaml

setup: |
  echo hi

run: |
  # Show workdir
  ls -l .
  # Show private/public storage copy/mount
  ls -l /s3-sky-managed-copy
  ls -l /s3-sky-managed-mount
  ls -l /s3-external-private-by-user
  ls -l /s3-external-public-by-user
  ls -l /s3-external-public-not-by-user
  ls -l /gcs-sky-managed-copy
  ls -l /gcs-sky-managed-mount
  ls -l /gcs-external-private-by-user
  ls -l /gcs-external-public-by-user
  ls -l /gcs-external-public-not-by-user
  ls -l /az-sky-managed-copy
  ls -l /az-sky-managed-mount
  ls -l /az-external-private-by-user
  ls -l /az-external-public-by-user
  ls -l /az-external-public-not-by-user
  # Write files on a mounted storage with access
  date > /s3-sky-managed-mount/hellotest.txt
  date > /s3-external-private-by-user/hellotest.txt
  date > /s3-external-public-by-user/hellotest.txt
  date > /gcs-sky-managed-mount/hellotest.txt
  date > /gcs-external-private-by-user/hellotest.txt
  date > /gcs-external-public-by-user/hellotest.txt
  date > /az-sky-managed-mount/hellotest.txt
  date > /az-external-private-by-user/hellotest.txt
  date > /az-external-public-by-user/hellotest.txt
  # Confirm file was written
  cat /s3-sky-managed-mount/hellotest.txt
  cat /s3-external-private-by-user/hellotest.txt
  cat /s3-external-public-by-user/hellotest.txt
  cat /gcs-sky-managed-mount/hellotest.txt
  cat /gcs-external-private-by-user/hellotest.txt
  cat /gcs-external-public-by-user/hellotest.txt
  cat /az-sky-managed-mount/hellotest.txt
  cat /az-external-private-by-user/hellotest.txt
  cat /az-external-public-by-user/hellotest.txt
```

### CODE REVIEW

**Comment by landscapepainter** [comment]
@cblmemo @romilbhardwaj @Michaelvll ran additional backwards compatibility test and noted at the PR description at the top. This is ready for a look!

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @landscapepainter

File: sky/data/storage.py [hunk 1]
```diff
@@ -1942,8 +1942,12 @@ class AzureBlobStore(AbstractStore):
     """Represents the backend for Azure Blob Storage Container."""
 
     _ACCESS_DENIED_MESSAGE = 'Access Denied'
-    DEFAULT_STORAGE_ACCOUNT_NAME = 'sky{region}{user_hash}'
     DEFAULT_RESOURCE_GROUP_NAME = 'sky{user_hash}'
+    # Unlike resource group names, which only need to be unique within the
+    # subscription, storage account names must be globally unique across all of
+    # Azure users. Hence, the storage account name includes the subscription
+    # hash as well to ensure its uniqueness.
+    DEFAULT_STORAGE_ACCOUNT_NAME = 'sky{region}{user_hash}{subscription_hash}'
```
[the remark this answers is not in the document]
Comment (landscapepainter) [reply sky/data/storage.py]: Thanks for catching this. We indeed have to shorten this and I hashed the region as well and kept it to have length of 4. So this default storage account name will have a fixed length of 15 characters now.

**Comment by landscapepainter** [comment]
@romilbhardwaj This is ready for another look!

**Review by romilbhardwaj — APPROVED** [review]
Thanks @landscapepainter. LGTM if tests pass.

**Pull request merged**